cmd/osquery-perf/agent.go

```diff
@@ -1242,6 +1261,135 @@ func (a *agent) execScripts(execIDs []string, orbitClient *service.OrbitClient)
+			a.installSoftwareItem(installerID, orbitClient)
+		}
+	}
+}
+
+func (a *agent) installSoftwareItem(installerID string, orbitClient *service.OrbitClient) {
+	payload := &fleet.HostSoftwareInstallResultPayload{}
+	payload.InstallUUID = installerID
+	installer, err := orbitClient.GetInstallerDetails(installerID)
+	if err != nil {
+		log.Println("get installer details:", err)
+		return
+	}
+	failed := false
+	if installer.PreInstallCondition != "" {
+		time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
+		if installer.PreInstallCondition == "select 1" {
+			// Always pass
+			payload.PreInstallConditionOutput = ptr.String("1")
+		} else if installer.PreInstallCondition == "select 0" ||
+			a.softwareInstaller.preInstallFailureProb > 0.0 && rand.Float64() <= a.softwareInstaller.preInstallFailureProb {
+			// Fail
+			payload.PreInstallConditionOutput = ptr.String("")
+			failed = true
+		} else {
+			payload.PreInstallConditionOutput = ptr.String("1")
+		}
+	}
+
+	var path string
+	if !failed {
+		tmpDir, err := os.MkdirTemp("", "")
+		if err != nil {
+			log.Println("create temp dir:", err)
+			return
+		}
+		defer os.RemoveAll(tmpDir)
+		path, err = orbitClient.DownloadSoftwareInstaller(installer.InstallerID, tmpDir)
+		if err != nil {
+			log.Println("download software installer:", err)
+			return
+		}
+
+		time.Sleep(time.Duration(rand.Intn(30)) * time.Second)
+		if installer.InstallScript == "exit 0" {
+			// Always pass
+			payload.InstallScriptExitCode = ptr.Int(0)
+			payload.InstallScriptOutput = ptr.String("Installed on osquery-perf (always pass)")
+		} else if installer.InstallScript == "exit 1" {
+			payload.InstallScriptExitCode = ptr.Int(1)
+			payload.InstallScriptOutput = ptr.String("Installed on osquery-perf (always fail)")
+			failed = true
+		} else if a.softwareInstaller.installFailureProb > 0.0 && rand.Float64() <= a.softwareInstaller.installFailureProb {
+			payload.InstallScriptExitCode = ptr.Int(1)
+			payload.InstallScriptOutput = ptr.String("Installed on osquery-perf (fail)")
+			failed = true
+		} else {
+			payload.InstallScriptExitCode = ptr.Int(0)
+			payload.InstallScriptOutput = ptr.String("Installed on osquery-perf (pass)")
+		}
+	}
+	if !failed {
+		// Figure out what we're actually installing here and add it to software inventory
+		f, err := os.Open(path)
+		if err != nil {
+			log.Println("open installer:", err)
+			return
+		}
+		defer f.Close()
+		meta, err := file.ExtractInstallerMetadata(f)
+		if err != nil {
+			log.Println("extract installer metadata:", err)
+			return
+		}
+		key := meta.Name + "+" + meta.Version + "+" + meta.BundleIdentifier
+		if a.installedSoftware == nil {
+			a.installedSoftware = make(map[string]map[string]string)
+		}
+		source := ""
+		switch a.os {
+		case "macos":
+			source = "apps"
+		case "windows":
+			source = "programs"
+		case "ubuntu":
+			source = "deb_packages"
+		default:
+			log.Printf("unknown OS to software installer: %s", a.os)
+			return
+		}
```

Comment:
Check if `key` is already in `a.installedSoftware`?

Reply:
Yes, good catch